### PR TITLE
Bug 1225086 - [TV][2.5] Marketplace can add or delete app in contextmenu

### PR DESCRIPTION
**Pull request by rickychien**

### CODE REVIEW

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=e80d83552daa4713cf9cfde75db70bf48fc9f64c

**Comment by autolander** [comment]
Autolander could not find a bug number in your pull request title. All pull requests should be in the format of: Bug [number] - [description].

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=305fa95ccaa9c0bd1ee3aa8d840edb7e70e2155f

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=11cf23a8b5243412dbda6fc281f53f6aad8130a7

File: tv_apps/smart-system/js/app_install_manager.js [hunk 1]
```diff
@@ -102,10 +101,11 @@ var AppInstallManager = {
   },
 
   isMarketplaceAppActive: function ai_isMarketplaceAppActive() {
-    const MARKETPLACE_DOMAIN = 'https://marketplace.firefox.com/';
+    const MARKETPLACE_DOMAIN = 'marketplace';
```
Comment (rickychien) [inline tv_apps/smart-system/js/app_install_manager.js]: We'll remove it once patch gets r+

File: tv_apps/smart-system/js/app_install_manager.js [hunk 2]
```diff
@@ -102,10 +101,11 @@ var AppInstallManager = {
...
 
     var activeApp = AppWindowManager.getActiveApp();
-    return activeApp && activeApp.manifestURL.startsWith(MARKETPLACE_DOMAIN);
+    return activeApp &&
+      activeApp.manifestURL.indexOf(MARKETPLACE_DOMAIN) !== -1;
```
Comment (rickychien) [inline tv_apps/smart-system/js/app_install_manager.js]: same as above

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=c0b51049668eb94b9d08b864b084d687f4220e85

File: tv_apps/smart-system/js/app_install_manager.js [hunk 3]
```diff
@@ -384,11 +391,13 @@ var AppInstallManager = {
       return;
     }
 
-    if (app.manifestURL === this.getPreviewAppManifestURL()) {
+    if (this.enablePreview) {
```
Comment (luke-chang) [inline tv_apps/smart-system/js/app_install_manager.js]: You still need to check if the app needs to be previewed or not.

File: tv_apps/smart-system/js/browser_context_menu.js [hunk 4]
```diff
@@ -204,14 +206,37 @@
     // context menu api
     if (detail.contextmenu && detail.contextmenu.items.length) {
       var that = this;
-      detail.contextmenu.items.forEach(function(choice, index) {
+      detail.contextmenu.items.forEach(function(choice) {
+        var label;
+        var clickAction;
+
+        if (AppInstallManager.isMarketplaceAppActive()) {
```
Comment (luke-chang) [inline tv_apps/smart-system/js/browser_context_menu.js]: - Put this logic to AppInstallManager via CustomEvent.
- Stop handling contextmenu here.

File: tv_apps/smart-system/js/app_install_manager.js [hunk 5]
```diff
@@ -373,6 +375,11 @@ var AppInstallManager = {
     }
   },
 
+  handleAddApp: function(app) {
```
Comment (luke-chang) [inline tv_apps/smart-system/js/app_install_manager.js]: Never used in this patch

File: tv_apps/smart-system/js/browser_context_menu.js (hunk 4 above)
Comment (rickychien) [inline tv_apps/smart-system/js/browser_context_menu.js]: I was still thinking some of sections are related to app_install_manager but some of them are related to browser_context_menu ex: label & clickAction which should belong to browser_context_menu.
IMO, I would like to leave it here

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=c2cd13145266ee48be406294910369f8cbb5588b

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=8273be6d6a465ae553237b1f7a3489e8c6ede727

File: tv_apps/smart-system/js/app_install_manager.js [hunk 6]
```diff
@@ -45,6 +45,7 @@ var AppInstallManager = {
     this.appInfos = {};
     this.setupQueue = [];
     this.isSetupInProgress = false;
+    this.enablePreview = true;
```
Comment (luke-chang) [inline tv_apps/smart-system/js/app_install_manager.js]: Rename it to something like `pausePreview`

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=b784df0fec4cc386ff30d62544ffd4b38329d6f0

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=37ac472eda8e692fae53bf6a99ead63962105984

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=b52060ee2ccdf259c13b4b9dca66548b992d6306

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=1d69be24247a39a64579ab91388327bfd6828e63

File: tv_apps/smart-system/js/browser_context_menu.js [hunk 7]
```diff
@@ -204,14 +206,46 @@
...
+          } else {
+            textL10nId = 'delete-from-apps';
+            icon = 'style/icons/default.png';
+            clickAction = function _clickAction() {
+              navigator.mozApps.mgmt.uninstall(app);
+            };
+          }
+        } else {
+          textRaw = choice.label;
+          clickAction = function _clickAction() {
```
Comment (luke-chang) [inline tv_apps/smart-system/js/browser_context_menu.js]: you miss the `icon` here!

File: tv_apps/smart-system/js/preview_window.js [hunk 8]
```diff
@@ -118,16 +115,16 @@
...
         type: 'menu',
         customized: true,
         items: [{
           icon: icon,
-          id: id,
           label: value,
           type: 'menuitem',
-          onClick: onClick
+          onClick: onClick,
+          usePreviewContextMenu: true
```
Comment (luke-chang) [inline tv_apps/smart-system/js/preview_window.js]: remove this line.

File: tv_apps/smart-system/js/browser_context_menu.js [hunk 9]
```diff
@@ -204,14 +206,46 @@
...
+            };
+          } else {
+            textL10nId = 'delete-from-apps';
+            icon = 'style/icons/default.png';
+            clickAction = function _clickAction() {
+              navigator.mozApps.mgmt.uninstall(app);
+            };
+          }
+        } else {
+          textRaw = choice.label;
```
Comment (luke-chang) [inline tv_apps/smart-system/js/browser_context_menu.js]: you miss `icon = choice.icon` here!

**Comment by autolander** [comment]
Running tests
https://treeherder.mozilla.org/#/jobs?repo=gaia&revision=95c64d15865b6c715efb0463b34f64c524739293

**Pull request merged**